File: app/Policies/StudyGroupPolicy.php
<?php

namespace App\Policies;

use App\Models\StudyGroup;
use App\Models\StudyGroupMember;
use App\Models\User;

class StudyGroupPolicy
{
    public function view(User $user, StudyGroup $group): bool
    {
        if ($group->visibility === StudyGroup::VISIBILITY_PUBLIC) {
            return true;
        }

        return $this->isMember($user, $group) || $user->isCircleMemberWith($group->owner);
    }

    public function update(User $user, StudyGroup $group): bool
    {
        return $this->memberRole($user, $group) === StudyGroupMember::ROLE_OWNER;
    }

    public function join(User $user, StudyGroup $group): bool
    {
        if (! $group->isActive()) {
            return false;
        }

        if ($this->isMember($user, $group)) {
            return true;
        }

        return match ($group->visibility) {
            StudyGroup::VISIBILITY_PUBLIC, StudyGroup::VISIBILITY_PASSWORD => true,
            StudyGroup::VISIBILITY_FRIENDS => $user->isCircleMemberWith($group->owner),
            default => false,
        };
    }

    public function leave(User $user, StudyGroup $group): bool
    {
        return $this->isMember($user, $group)
            && $this->memberRole($user, $group) !== StudyGroupMember::ROLE_OWNER;
    }

    public function manageFocusRooms(User $user, StudyGroup $group): bool
    {
        return in_array($this->memberRole($user, $group), [
            StudyGroupMember::ROLE_OWNER,
            StudyGroupMember::ROLE_ADMIN,
        ], true);
    }

    public function startFocusStudy(User $user, StudyGroup $group): bool
    {
        return $group->isActive() && $this->isMember($user, $group);
    }

    private function isMember(User $user, StudyGroup $group): bool
    {
        return $group->members()
            ->where('user_id', $user->id)
            ->exists();
    }

    private function memberRole(User $user, StudyGroup $group): ?string
    {
        return $group->members()
            ->where('user_id', $user->id)
            ->value('role');
    }
}

File: app/policies/study_group_policy.rb
class StudyGroupPolicy < ApplicationPolicy
  def show?
    view?
  end

  def view?
    record.visibility == StudyGroup::VISIBILITY_PUBLIC || member? || user&.is_circle_member_with?(record.owner)
  end

  def update?
    member_role == StudyGroupMember::ROLE_OWNER
  end

  def join?
    return false unless record.active?
    return true if member?

    case record.visibility
    when StudyGroup::VISIBILITY_PUBLIC, StudyGroup::VISIBILITY_PASSWORD then true
    when StudyGroup::VISIBILITY_FRIENDS then user&.is_circle_member_with?(record.owner)
    else false
    end
  end

  def leave?
    member? && member_role != StudyGroupMember::ROLE_OWNER
  end

  def manage_focus_rooms?
    [StudyGroupMember::ROLE_OWNER, StudyGroupMember::ROLE_ADMIN].include?(member_role)
  end

  def start_focus_study?
    record.active? && member?
  end

  private

  def member?
    record.members.exists?(user_id: user&.id)
  end

  def member_role
    record.members.where(user_id: user&.id).pick(:role)
  end
end
